extensions/ql-vscode/src/view/results/raw-results-table.tsx

```diff
@@ -19,158 +18,157 @@ import { onNavigation } from "./results";
 import { tryGetResolvableLocation } from "../../common/bqrs-utils";
 import { ScrollIntoViewHelper } from "./scroll-into-view-helper";
 import { sendTelemetry } from "../common/telemetry";
+import { assertNever } from "../../common/helpers-pure";
 
-export type RawTableProps = ResultTableProps & {
+export type RawTableProps = {
+  databaseUri: string;
   resultSet: RawTableResultSet;
   sortState?: RawResultsSortState;
   offset: number;
 };
 
-interface RawTableState {
-  selectedItem?: { row: number; column: number };
+interface TableItem {
+  readonly row: number;
+  readonly column: number;
 }
 
-export class RawTable extends React.Component<RawTableProps, RawTableState> {
-  private scroller = new ScrollIntoViewHelper();
-
-  constructor(props: RawTableProps) {
-    super(props);
-    this.setSelection = this.setSelection.bind(this);
-    this.handleNavigationEvent = this.handleNavigationEvent.bind(this);
-    this.state = {};
-  }
-
-  private setSelection(row: number, column: number) {
-    this.setState((prev) => ({
-      ...prev,
-      selectedItem: { row, column },
-    }));
-    sendTelemetry("local-results-raw-results-table-selected");
-  }
-
-  render(): React.ReactNode {
-    const { resultSet, databaseUri } = this.props;
-
-    let dataRows = resultSet.rows;
-    if (dataRows.length === 0) {
-      return emptyQueryResultsMessage();
-    }
-
-    let numTruncatedResults = 0;
-    if (dataRows.length > RAW_RESULTS_LIMIT) {
-      numTruncatedResults = dataRows.length - RAW_RESULTS_LIMIT;
-      dataRows = dataRows.slice(0, RAW_RESULTS_LIMIT);
-    }
-
-    const tableRows = dataRows.map((row: ResultRow, rowIndex: number) => (
-      <RawTableRow
-        key={rowIndex}
-        rowIndex={rowIndex + this.props.offset}
-        row={row}
-        databaseUri={databaseUri}
-        selectedColumn={
-          this.state.selectedItem?.row === rowIndex
-            ? this.state.selectedItem?.column
-            : undefined
+export function RawTable({
+  databaseUri,
+  resultSet,
+  sortState,
+  offset,
+}: RawTableProps) {
+  const [selectedItem, setSelectedItem] = React.useState<
+    TableItem | undefined
+  >();
+
+  const scroller = React.useMemo(() => new ScrollIntoViewHelper(), []);
+  scroller.update();
```

Comment:
I think the behaviour we want for the scroller is we want one scroller per `RawTable` instance, and we don't want to create a new one every time we render. So I'm using `useMemo` to cache and instance but then calling `scroller.update()` directly every time we render. Is this ok? It seems to work. Internally the scroller keeps a `shouldScrollIntoView` field so it'll only actually do anything when it has previously been instructed to do so.

Since the `ScrollIntoViewHelper` class is used from another location I don't want to go too hard on refactoring it right now, but once we've also dealt with `PathTable` we could update the scroller to fit with a functional component better, such as ditching `shouldScrollIntoView` and being triggered using a `useEffect` hook on `selectedItem`.

Reply:
I think it would be slightly better to call `useEffect` here to make it explicit that we're calling this method on every render. `useEffect` without any dependencies is run on every render as well, but I think it will check if a re-render is necessary after it, so it should be more correct:

```suggestion
  useEffect(() => scroller.update());
```

Reply:
Ah, I had forgotten that `useEffect` without any second argument at all (or I guess implicitly passing `undefined`) had that behaviour. I agree that would be better.

I remember always feeling like the empty array and undefined should be the same here, but actually there's complete opposite behaviours. Just got to remember that.